README.md
# qiClock


Changelog.txt
todo: Add MapCoords to enemy, marked and shortkey

/script DEFAULT_CHAT_FRAME:AddMessage( GetMouseFocus():GetName() );

SendAddonMessage("prefix", "text", "type", "target");
local nameplateUnit = C_NamePlate.GetNamePlateForUnit("target");

for i=1,40 do
    local name, icon, _, _, _, etime = UnitBuff("player",i)
    if name == "Summon Chauffeur" or name == "Ban-Lu, Grandmaster's Companion" then
        --do things
    end
end



TODO:

- DPS counter, perhaps like FPS viewer?
- Vendor price in GameTooltip
- Used in recepies or in quests in items.
- played today
- Multibletooltip
- tooltip notes
- thanks for buff
- who on whisper
- marktool baudmark
- auto dismount
- extended map info.

- More sub catagories in addon settings.


init.lua
local name, methods = ...


-- Global Title and version from toc file
methods.title = "qiClock: "
methods.eventFrame = CreateFrame("Frame", "eventFrame")
methods.eventFrame:RegisterEvent("ADDON_LOADED")
methods.eventFrame:RegisterEvent("PLAYER_LOGIN")
methods.eventFrame:RegisterEvent("PLAYER_TARGET_CHANGED")
--methods.eventFrame:RegisterEvent("UNIT_AURA")

methods.eventFrame:RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED")

-- Call right event
methods.eventFrame:SetScript("OnEvent",
    function(self, event, ...)
        if (type(methods[event]) == 'function') then
            methods[event](self, event, ...);
        end
    end
)


if not qvars then qvars = {} end

function methods:init_vars()
    methods.units = {}

    methods.ICAuras = {
        {id = 8177, cooldown = 0.05, duration = 45, type = "grnd totem"},
        {id = 8143, cooldown = 0.05, duration = 120, type = "trem totem"},
        {id = 7744, cooldown = 115, duration = 5, type = "WOTF"},
        {id = 45438, cooldown = 290, duration = 10, type = "Ice Block"},
        {id = 27148, cooldown = 0.05, duration = 30, type = "Void sacrifice"},
        {id = 25771, cooldown = 120, duration = 0.01, type = "Pally Bubbles"},
        {id = 498, cooldown = 180, duration = 12, type = "Div Prot"},
        {id = 642, cooldown = 300, duration = 12, type = "Divine Shield"},
        {id = 18499, cooldown = 30, duration = 10, type = "Bers Rage"},
        {id = 34471, cooldown = 120, duration = 18, type = "Beast Within"},
        {id = 31224, cooldown = 60, duration = 5, type = "Cloak of Shad"},
        {id = 6346, cooldown = 180, duration = 180, type = "Fear Ward"},
        {id = 59752, cooldown = 120, duration = 0.01, type = "Every Man"},
        {id = 23920, cooldown = 10, duration = 5, type = "Spell Reflection"}
    };

    methods.ICFears = {
        id = 5782, --Fear
        id = 6213,
        id = 6215,
        id = 5484, -- Howl of Terror
        id = 17928,
        id = 6358, -- Seduction
        id = 8122, -- Psychic Scream
        id = 8124,
        id = 10888,
        id = 10890,
        id = 5246, -- Intimidating Shout
        id = 16377, -- transform (caused by polymorph / hex)
        id = 16372,
        id = 32570,
        id = 16371,
        id = 32789,
        id = 2914,
        id = 32822,
        id = 13321
    }
end

methods:init_vars()


qiFrames.lua

local name, methods = ...
-- Create a list with unit and spells (5)

-- Create Frame
methods.ImmunityClockTargetFrame = CreateFrame("Frame", "ImmunityClockTargetFrame", UIParent)
-- ImmunityClockTargetFrame:SetTopLevel(true)
methods.ImmunityClockTargetFrame:SetFrameStrata("BACKGROUND")
methods.ImmunityClockTargetFrame:SetMovable(true)
methods.ImmunityClockTargetFrame:SetWidth(128)
methods.ImmunityClockTargetFrame:SetHeight(32)
methods.ImmunityClockTargetFrame:SetPoint("BOTTOMLEFT", "TargetFrameHealthBar", "BOTTOMLEFT", -2, -35);
methods.ImmunityClockTargetFrame:EnableMouse(true)
methods.ImmunityClockTargetFrame:RegisterForDrag("LeftButton")

-- Set Background texture
-- local t = ImmunityClockTargetFrame:CreateTexture(nil, "BACKGROUND")
-- t:SetTexture("Interface\\DialogFrame\\UI-DialogBox-Background")
-- t:SetAllPoints(ImmunityClockTargetFrame, - 1, - 1)
-- t:SetVertexColor(1, 1, 1, .5)
-- ImmunityClockTargetFrame.texture = t

-- Create a fontframe1 -- Need 3 font strings
methods.ImmunityClockTargetFrame.title1 = ImmunityClockTargetFrame:CreateFontString(nil, "OVERLAY")
methods.ImmunityClockTargetFrame.title1:SetFont(STANDARD_TEXT_FONT, 12, "OUTLINE")
methods.ImmunityClockTargetFrame.title1:SetTextColor(1, 1, 0, .8)
methods.ImmunityClockTargetFrame.title1:SetPoint("TOP", ImmunityClockTargetFrame, "TOP", 0, -35)
methods.ImmunityClockTargetFrame.title1:SetText("Line 1")

methods.ImmunityClockTargetFrame.title2 = ImmunityClockTargetFrame:CreateFontString(nil, "OVERLAY")
methods.ImmunityClockTargetFrame.title2:SetFont(STANDARD_TEXT_FONT, 12, "OUTLINE")
methods.ImmunityClockTargetFrame.title2:SetTextColor(1, 1, 0, .8)
methods.ImmunityClockTargetFrame.title2:SetPoint("TOP", ImmunityClockTargetFrame, "TOP", 0, -49)
methods.ImmunityClockTargetFrame.title2:SetText("Line 2")

methods.ImmunityClockTargetFrame.title3 = ImmunityClockTargetFrame:CreateFontString(nil, "OVERLAY")
methods.ImmunityClockTargetFrame.title3:SetFont(STANDARD_TEXT_FONT, 12, "OUTLINE")
methods.ImmunityClockTargetFrame.title3:SetTextColor(1, 1, 0, .8)
methods.ImmunityClockTargetFrame.title3:SetPoint("TOP", ImmunityClockTargetFrame, "TOP", 0, -63)
methods.ImmunityClockTargetFrame.title3:SetText("Line 3")

-- Scripts
methods.ImmunityClockTargetFrame:SetScript("OnUpdate", function(self)

end)

methods.ImmunityClockTargetFrame:SetScript("OnDragStart", ImmunityClockTargetFrame.StartMoving)
methods.ImmunityClockTargetFrame:SetScript("OnDragStop", ImmunityClockTargetFrame.StopMovingOrSizing)
methods.ImmunityClockTargetFrame:Hide()
--[[


-- Clock frame
ImmunityClockWindowFrame = CreateFrame("Frame", "ImmunityClockWindowFrame", UIParent)
ImmunityClockWindowFrame:SetWidth(250)
ImmunityClockWindowFrame:SetHeight(96)
ImmunityClockWindowFrame:SetFrameStrata("BACKGROUND")
ImmunityClockWindowFrame:EnableMouse(true)
ImmunityClockWindowFrame:SetMovable(true)
ImmunityClockWindowFrame:RegisterForDrag("LeftButton")
ImmunityClockWindowFrame:SetClampedToScreen(true)
ImmunityClockWindowFrame:SetPoint("CENTER", 0, 0);

-- Set Background texture
local t = ImmunityClockWindowFrame:CreateTexture(nil, "BACKGROUND")
t:SetTexture("Interface\\DialogFrame\\UI-DialogBox-Background")
t:SetAllPoints(ImmunityClockWindowFrame, - 1, - 1)
t:SetVertexColor(1, 1, 1, .5)
ImmunityClockWindowFrame.texture = t

-- The edge graphics
ImmunityClockWindowFrame:SetBackdrop( {
bgFile = "Interface/DialogFrame/UI-DialogBox-Background",
edgeFile = "Interface/Tooltips/UI-Tooltip-Border", tile = true, tileSize = 16, edgeSize = 16,
insets = { left = 5, right = 5, top = 5, bottom = 5 }
});

-- Title
local title = ImmunityClockWindowFrame:CreateTexture("FOREGROUND")
title:SetPoint("TOP", ImmunityClockWindowFrame, "TOP", 0, -5)
title:SetHeight(12)
title:SetWidth(250)

title.title = ImmunityClockWindowFrame:CreateFontString(nil, "OVERLAY")
title.title:SetFont(STANDARD_TEXT_FONT, 12, "OUTLINE")
title.title:SetTextColor(1, 1, 0, .8)
title.title:SetPoint("TOP", title, "TOP")
title.title:SetText("Immunity Clock")

-- Line 1
local text1 = ImmunityClockWindowFrame:CreateTexture("ARTWORK")
text1:SetPoint("TOP", title, "TOP", 0, 0)
text1:SetHeight(12)
text1:SetWidth(250)

text1.text1 = ImmunityClockWindowFrame:CreateFontString("ImmunityClockWindowText1")
text1.text1:SetFont(STANDARD_TEXT_FONT, 12, "OUTLINE")
text1.text1:SetPoint("TOPLEFT", text1, "BOTTOMLEFT", 10, 0)
text1.text1:SetText("Line 1")

-- Line 2
local text2 = ImmunityClockWindowFrame:CreateTexture("ARTWORK")
text2:SetPoint("TOP", text1, "TOP", 0, 0)
text2:SetHeight(12)
text2:SetWidth(250)

text2.text2 = ImmunityClockWindowFrame:CreateFontString("ImmunityClockWindowText1")
text2.text2:SetFont(STANDARD_TEXT_FONT, 12, "OUTLINE")
text2.text2:SetPoint("TOPLEFT", text2, "BOTTOMLEFT", 10, -14)
text2.text2:SetText("Line 2")

-- Line 3
local text3 = ImmunityClockWindowFrame:CreateTexture("ARTWORK")
text3:SetPoint("TOP", text2, "TOP", 0, 0)
text3:SetHeight(12)
text3:SetWidth(250)

text3.text3 = ImmunityClockWindowFrame:CreateFontString("ImmunityClockWindowText1")
text3.text3:SetFont(STANDARD_TEXT_FONT, 12, "OUTLINE")
text3.text3:SetPoint("TOPLEFT", text2, "BOTTOMLEFT", 10, -28)
text3.text3:SetText("Line 3")

-- Line 4
local text4 = ImmunityClockWindowFrame:CreateTexture("ARTWORK")
text4:SetPoint("TOP", text3, "TOP", 0, 0)
text4:SetHeight(12)
text4:SetWidth(250)

text4.text4 = ImmunityClockWindowFrame:CreateFontString("ImmunityClockWindowText1")
text4.text4:SetFont(STANDARD_TEXT_FONT, 12, "OUTLINE")
text4.text4:SetPoint("TOPLEFT", text4, "BOTTOMLEFT", 10, -42)
text4.text4:SetText("Line 4")

-- Line 5
local text5 = ImmunityClockWindowFrame:CreateTexture("ARTWORK")
text5:SetPoint("TOP", text4, "TOP", 0, 0)
text5:SetHeight(12)
text5:SetWidth(250)

text5.text5 = ImmunityClockWindowFrame:CreateFontString("ImmunityClockWindowText1")
text5.text5:SetFont(STANDARD_TEXT_FONT, 12, "OUTLINE")
text5.text5:SetPoint("TOPLEFT", text5, "BOTTOMLEFT", 10, -56)
text5.text5:SetText("Line 5")

-- Close button
local button = CreateFrame("Button", "ImmunityClockWindowFrame")
button:SetWidth(24)
button:SetHeight(24)
button:SetPoint("TOPRIGHT", "ImmunityClockWindowFrame", 2, -2)
button:SetNormalTexture("Interface/Buttons/UI-Panel-MinimizeButton-Up")
button:SetPushedTexture("Interface/Buttons/UI-Panel-MinimizeButton-Down")
button:SetHighlightTexture("Interface/Buttons/UI-Panel-MinimizeButton-Highlight")

button:SetScript("OnClick", function()
print("click")
end)

ImmunityClockWindowFrame.button = button

ImmunityClockWindowFrame:SetScript("OnLoad", function()

end)

ImmunityClockWindowFrame:SetScript("OnDragStart", ImmunityClockWindowFrame.StartMoving)
ImmunityClockWindowFrame:SetScript("OnDragStop", ImmunityClockWindowFrame.StopMovingOrSizing)

ImmunityClockWindowFrame:Show()
--]]


qiClockEventHandler.lua
-- Use wow own table, add line belove in all files to make them 'localy global'
local name, methods = ...

-- check for 'fear' debuff
-- check for 'fear' antidote
-- start timer on antidotes/ fear
-- Show icon with counter or something like it

-- Tru to brake this down into smaller functions
-- make use of lua classes
-- test libAurora

-- Event methods, events are registred in init
function methods:ADDON_LOADED(self, event)
  if event == "qiClock" then
    print("qiClock loaded");
  end
end

function methods:PLAYER_LOGIN(self, event)

end

function methods:PLAYER_TARGET_CHANGED(self, arg1)


  if _G["TargetFrameHealthBar"]:IsVisible() then
    methods.ImmunityClockTargetFrame:Show()
  else
    methods.ImmunityClockTargetFrame:Hide()
  end

  local uGuid, uName = UnitGUID("target"), UnitName("target")

  -- This only works for enemy if, detect magic is used
  -- if combat log haven't record this yet, we get an error
  -- methods:checkTimeOut(uGuid)
  methods:AddSpellsToTable(spellName, playerName, uName, uGuid)
  gotBuffs, buffs = methods:checkBuff("target")

  if methods.units[uGuid] then
    methods:guidHasBuffs(uGuid, uName)
  end
end

function methods:COMBAT_LOG_EVENT_UNFILTERED(self, event)
  -- if event == nil then return end
  local playerGuid = UnitGUID("player")
  local playerName = UnitName("player")
  -- Just remove the uGuid that this fire
  local timestamp, subevent, hideCaster, sourceGUID, sourceName, sourceFlags, sourceRaidFlags, destGUID, destName, destFlags, destRaidFlags = CombatLogGetCurrentEventInfo()
  local spellId, spellName, spellSchool = select(12, CombatLogGetCurrentEventInfo())
  local amount, overkill, school, resisted, blocked, absorbed, critical, glancing, crushing, isOffHand = select(15, CombatLogGetCurrentEventInfo())

  -- We can't get spellID from combat log or UNIT_AURA
  -- only spellNames. If someone uses Detect magic we can read more on target.

  -- sourceName is always the caster (for all)
  -- destName is always the reciever (for all)

  if subevent == "SPELL_CAST_START" and playerGuid == sourceGUID then
    --  print(playerName .. " start casting " .. spellName)
  end

  if subevent == "SPELL_CAST_SUCCESS" and playerGuid == sourceGUID then
    --print(playerName .. " casts " .. spellName)
  end

  if subevent == "SPELL_CAST_FAILED" and playerGuid == sourceGUID then
    --print(playerName .. " failed casting " .. spellName)
  end

  if subevent == "SPELL_AURA_APPLIED" then
    --methods:checkTimeOut(sourceGUID)
    methods:AddSpellsToTable(spellName, sourceName, sourceGUID)
  end

  if subevent == "SPELL_AURA_REMOVED" then
    if not sourceName then return end

    --  if HostilePlayer(sourceFlags) then
    if methods.units[sourceGUID] then

      if methods.units[sourceGUID][spellName]['spellStartTime'] then
        local start = methods.units[sourceGUID][spellName]['spellStartTime']
        local duration = GetServerTime() - start

        if methods.units[sourceGUID][spellName]['duration'] and
        methods.units[sourceGUID][spellName]['duration'] < duration then
          methods.units[sourceGUID][spellName]['duration'] = duration
        end
      end

      methods.units[sourceGUID][spellName]['spellActive'] = false
      methods.units[sourceGUID][spellName]['spellStartTime'] = nil
      methods.units[sourceGUID][spellName]['cooldown'] = nil

      print(sourceName .. " removed '" .. spellName .. "'")
      --  end
    end
  end
end

function HostilePlayer(sourceFlags)
  if bit.band(sourceFlags, COMBATLOG_OBJECT_CONTROL_PLAYER) > 0 then -- check if player controlled
    --  bit.band(sourceFlags, COMBATLOG_OBJECT_REACTION_HOSTILE) > 0 then
    return true
  end

  return false
end


function methods:UNIT_AURA(self, arg1)
  -- Works primaly with unit like player, target, raid1-40 party 1-5
  local uGuid, uName = UnitGUID(arg1), UnitName(arg1)

  if uGuid then
    -- local type, zero, server_id, instance_id, zone_uid, npc_id, spawn_uid = strsplit("-", uGuid);
    methods:checkBuff(arg1)
  end
end

function methods:checkBuff(arg1)
  local sourceGUID = UnitGUID(arg1)
  local sourceName = UnitName(arg1)

  -- We can't get buffs from enemy player, without detecting magic.
  for index = 1, 40 do
    local spellName, icon, _, _, duration, expirationTime, source, _, _, spellId = UnitBuff(arg1, index)
    if not spellName then return end

  --  methods:checkTimeOut(sourceGUID)
    methods:AddSpellsToTable(spellName, sourceName, sourceGUID, spellId, duration)
  end
end

function methods:has_value(tab, val)
  for index, value in ipairs(tab) do
    if value['id'] == val then
      return true
    end
  end

  return false
end


function methods:guidHasBuffs(uGuid, uName)
  -- Prints known buffs for unit.
  -- Add some sort of time out for spells
  -- View how long it been running.
  -- lot of false positive based on:
  -- combat log spamming same spell
  -- you can only track spelles you havr rekorded
  -- language based since we are forced to use spellName and not spellId
  -- spamming can accure then player moving in and out of range.

  if not methods.units[uGuid] then return end
  print()
  print(uName .. " BUFFS")

  for spellName, table in pairs(methods.units[uGuid]) do
    local active = methods.units[uGuid][spellName]['spellActive']
    local hardLimitTime = methods.units[uGuid][spellName]['hardLimitTime']
    local name = tostring(methods.units[uGuid][spellName]['playerName'])
    local spellId = methods.units[uGuid][spellName]['spellId']

    if not spellId then
      if active then
        print("    |cFF00FF00" .. spellName .. " " .. hardLimitTime .. "|r")
      else
        print("    |cFFFF0000" .. spellName .. " " .. hardLimitTime .. "|r")
      end
    elseif spellId then
      -- Party members, pet and yourself.
      local spell = GetSpellInfo(spellId)
      local rank = GetSpellSubtext(spellId) or ""

      if not rank then
        rank = ""
      end
      if active then
        print("    |cFF00FF00" .. spell .. " " .. rank .. "|r ")
      else
        print("    |cFFFF0000" .. spell .. " " .. rank .. "|r ")
      end
    end
  end
end

function methods:splitGuid(uGuid)
  -- local type, zero, server_id, instance_id, zone_uid, npc_id, spawn_uid = strsplit("-", uGuid);

  return strsplit("-", uGuid)
end

function methods:checkTimeOut(uGuid)
  if not methods.units[uGuid] then return end
  if not type(methods.units[uGuid]) == "table" then return end
  -- if not table.getn(methods.units[uGuid]) > 0 then return end

  for spellName, table in pairs(methods.units[uGuid]) do
    local start = methods.units[uGuid][spellName]['spellStartTime']
    local duration = methods.units[uGuid][spellName]['spellStartTime']
    local stop = start + duration
    local stopAll = start + methods.units[sourceGUID][spellName]['hardLimitTime']

    if getTime() > stop or getTime() > stopAll then
      methods.units[uGuid][spellName][spellActive] = false
      methods.units[uGuid][spellName]['spellStartTime'] = nil
    end
  end
end

function methods:AddSpellsToTable(spellName, sourceName, sourceGUID, spellId, duration)
  if not sourceName then return end
  -- if not HostilePlayer(sourceFlags) then return end

  if not methods.units[sourceGUID] then
    methods.units[sourceGUID] = {}
  end

  if not methods.units[sourceGUID][spellName] then
    methods.units[sourceGUID][spellName] = {}
  end

  if spellId then
    methods.units[sourceGUID][spellName]['spellId'] = spellId
  end

  if duration then
    methods.units[sourceGUID][spellName]['duration'] = duration
  end

  methods.units[sourceGUID][spellName]['spellStartTime'] = GetTime()
  methods.units[sourceGUID][spellName]['spellActive'] = true
  methods.units[sourceGUID][spellName]['playerName'] = sourceName
  methods.units[sourceGUID][spellName]['hardLimitTime'] = 180 -- add this from spelllist
  methods.units[sourceGUID][spellName]['cooldown'] = nil -- add this from spelllist

  print(sourceName .. " applied '" .. spellName .. "'")
end
